sql/stats: generation count was incorrectly incremented on each look up

fqazi opened the pull request:
Previously, an incorrect defer was causing the statistics generation to increment every time table statistics was looked up from the cache. This would cause the staleness optimization to kick in an inconsistent manner and be disabled needlessly. To address this, this patch only increments the generation updating stats.

This patch also contains an additional fix so that lease generations stay stable across long running queries. This was found because the test TestRenameTable started failing.

After this patch we see:
```
name                                  old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_only-10    2.97ms ±14%    2.69ms ± 7%  -9.43%  (p=0.008 n=9+10)

name                                  old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_only-10    1.14MB ± 1%    1.13MB ± 0%  -1.11%  (p=0.000 n=9+9)

name                                  old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_only-10     4.42k ± 2%     4.31k ± 0%  -2.55%  (p=0.000 n=9+9)
```

Fixes: #139876

Release note: None

fqazi committed:
sql/stats: generation count was incorrectly incremented on each look up

Previously, an incorrect defer was causing the statistics generation to
increment every time table statistics was looked up from the cache. This
would cause the staleness optimization to kick in an inconsistent
manner and be disabled needlessly. To address this, this patch only
increments the generation upon updates after the invalidation.

Fixes: #139876

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139877)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
Nice catch! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-:-OHd5WvE1MLh4W1KkPvu:b-qnzss8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-:-OHdJBHvDRJF0zd44pw9:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/stats/stats_cache.go` line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHdJ8rU0yO0oT7uEAB3:-OHdJ8rU0yO0oT7uEAB4:b8ejdrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3bf1920ed093e43c19c5eabc242afcd52d0f803/pkg/sql/stats/stats_cache.go#L348)):*
> ```Go
> 			// It will get re-incremented one more time when the
> 			// entry gets added back.
> 			sc.generation.Add(1)
> ```

Why do we need to increment it at all here?


<!-- Sent from Reviewable.io -->

fqazi committed:
catalog: ensure that lease generation stays fixed for long txns

Previously, for long running txn's the lease generation would always
get the latest value, which could then be set in the memo. This was
incorrect because a transaction at an older timestamp may intentionally
be observing the past (old descriptors), so the results of the staleness
check should not say the memo is valid for any later lease generations. This
would mean that tests with schema changes and long running txns like
TestTxnCanStillResolveOldName would fail, since the memo would be
incorrectly marked as not stale, because of a long running query. To
address this, this patch caches the lease generation inside the
descriptor collection at the start of any txn.

A simple example of this scenario:
conn 1:
BEGIN;
SELECT * FROM t;

conn 2:
ALTER TABLE t RENAME TO t2;

conn 1:
SELECT * FROM t;
COMMIT;
SELECT * FROM t; -- the memo staleness check should fail.

Release note: None

fqazi commented:
I had to add one more fix here to the lease generation stuff, since TestRenameTable started failing. For long running queries fetching the latest generation didn't work correctly, since those transactions observe the past. So, the second commit makes it so that the lease generation is cached at the start of any txn. An example of this:

```
    A simple example of this scenario:
    conn 1:
    BEGIN;
    SELECT * FROM t;

    conn 2:
    ALTER TABLE t RENAME TO t2;

    conn 1:
    SELECT * FROM t;
    COMMIT;
    SELECT * FROM t; -- the memo staleness check should fail.
```
@rafiss @mgartner Can you look at the second commit?

rafiss reviewed (changes requested):
it appears there's a testing issue
```
* SELECT 1
*
*
* ERROR: a panic has occurred!
* panic: unimplemented
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	pkg/sql/conn_executor.go:1007
*   | [...repeated from below...]
* Wraps: (2) while executing: SELECT _
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	pkg/sql/conn_executor.go:1007
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/kv.(*MockTransactionalSender).Epoch
*   | 	pkg/kv/mock_transactional_sender.go:218
*   | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Epoch
*   | 	pkg/kv/txn.go:277
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).serialize
*   | 	pkg/sql/conn_executor.go:4314
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run.func1
*   | 	pkg/sql/conn_executor.go:2299
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:261
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:881
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).GetLeaseGeneration
*   | 	pkg/sql/catalog/lease/lease.go:1785
*   | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetLeaseGeneration
*   | 	pkg/sql/catalog/descs/collection.go:224
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).maybeSetSQLLivenessSessionAndGeneration
*   | 	pkg/sql/conn_executor.go:4205
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper
*   | 	pkg/sql/conn_executor.go:4013
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-:-OHj6hjl9oq66CJ3rKXT:b-oo7h3d)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @fqazi, @mgartner, and @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 4205 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHj735Q3F5y8cLa1CWj:-OHj735Q3F5y8cLa1CWk:b-gy8udx) ([raw file](https://github.com/cockroachdb/cockroach/blob/49fb675ba3318de55d099f47c78a47aec9ba46a5/pkg/sql/conn_executor.go#L4205)):*
> ```Go
> 	}
> 	// Reset the lease generation at the same time.
> 	ex.extraTxnState.descCollection.GetLeaseGeneration()
> ```

could you explain further? how does calling `Get` cause a reset? looking at the implementation, it seems like it only does a "reset" if `ReleaseLeases` was already called earlier in the same transaction.

i wonder if we need other terminology instead of `Get` and `Reset` so this is more clear?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
@rafiss Test issue is fixed

> Reviewed 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-:-OHmRvBr9qTPgxAiEn71:bzaczog)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @rafiss, and @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 4205 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHj735Q3F5y8cLa1CWj:-OHmPWMZ9yfW4nxsCIis:b-f7872s) ([raw file](https://github.com/cockroachdb/cockroach/blob/49fb675ba3318de55d099f47c78a47aec9ba46a5/pkg/sql/conn_executor.go#L4205)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you explain further? how does calling `Get` cause a reset? looking at the implementation, it seems like it only does a "reset" if `ReleaseLeases` was already called earlier in the same transaction.

i wonder if we need other terminology instead of `Get` and `Reset` so this is more clear?
</blockquote></details>

Fair point, and I have renamed this to ResetLeaseGeneration. This is only called at the start of the txn now.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> Reviewed 1 of 4 files at r2, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-:-OHmWnOE5aht2WTMmnL2:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @fqazi and @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @fqazi and @rafiss)

___
*[`-- commits` line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHm_rfH7mmoYMxzgOZh:-OHm_rfH7mmoYMxzgOZi:btcwddd):*
nit: missing a word in "would always latest value"?
___
*[`pkg/sql/catalog/descs/factory.go` line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHmaKyr3l7CXrzc-itQ:-OHmaKyr3l7CXrzc-itR:bslm4nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1f8490595859314b25d8396133ac473811b3daf/pkg/sql/catalog/descs/factory.go#L153)):*
> ```Go
> 	if cf.leaseMgr == nil {
> 		lm = nil
> 	}
> ```

What is the reason for this `if` block? It looks like a no-op to me.
___
*[`pkg/sql/catalog/descs/collection.go` line 234 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHma7laDeuurx5fN3f1:-OHma7laDeuurx5fN3f2:b6wenzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1f8490595859314b25d8396133ac473811b3daf/pkg/sql/catalog/descs/collection.go#L234)):*
> ```Go
> // to make sure previously looked up information is still valid.
> func (tc *Collection) GetLeaseGeneration() int64 {
> 	// Return the cached lease generation, one should have been set earlier.g
> ```

nit: remove trailing "g"


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/catalog/descs/factory.go` line 153 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/139877#-OHmaKyr3l7CXrzc-itQ:-OHmtJkrDB-Dust-jW3L:b-tztcqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1f8490595859314b25d8396133ac473811b3daf/pkg/sql/catalog/descs/factory.go#L153)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What is the reason for this `if` block? It looks like a no-op to me.
</blockquote></details>

This is because we have an struct and an interface for LeaseManager. The collection takes the interface as input, and a nil struct pointer is still a non-nil interface (as weird as it sounds). So, if the collection does: l.lm != nil ...will be true even if the implementation *LeaseManager is nil.

Its because go allows you to do the following: https://go.dev/tour/methods/12


<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss @mgartner @yuzefovich TFTR

bors r+
